### ISSUE TITLE
SC2002: when cat is on a line on its own, suggest putting the redirection before the command instead

### ISSUE THREAD

**Issue by pabs3** [koalaman/shellcheck#1640]
#### For bugs
- Rule Id: SC2002
- My shellcheck version: 0.5.0

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/sh
cat file |
tr '[:cntrl:]' '?' |
sort |
uniq -c
```

#### Here's what shellcheck currently says:

```
In foo.sh line 2:
cat file |
    ^-- SC2002: Useless cat. Consider 'cmd < file | ..' or 'cmd file | ..' instead.
```

#### Here's what I wanted or expected to see:

```
In foo.sh line 2:
cat file |
tr '[:cntrl:]' '?' |
    ^-- SC2002: Useless cat. Consider this instead:
< file \
cmd |
```

#### Comments

Having the filename first is more aesthetically pleasing, especially with multi-line pipelines, which is why many people use cat, but they don't know about "< file cmd" redirections so then end up with useless-use-of-cat.

I'm not sure about the multi-line suggestion, but I think a quoted string would be more confusing due to the line continuation character and then escaped newline: "< file \\ncmd".

The rule's wiki page does put the filename before the command in one case, but doesn't have a multi-line case and the suggestions ShellCheck produces for SC2002 should reflect the multi-line case.

I found some [discussion](https://stackoverflow.com/questions/11710552/useless-use-of-cat) about this on StackOverflow.

I can't check shellcheck.net because it uses JavaScript.

**Comment by flexibeast** [comment]
In my testing, the idiom of putting the redirection first doesn't work in dash or bash when a `while` is involved, although it does work in zsh:

`test.data`:
```
a
b
c
```

`test.sh`:
```
#!/bin/dash
TEST=test.data
<"${TEST}" while read -r L
do
    echo "${L}"
done
```

dash output:
```
./test.sh: 4: while: not found
./test.sh: 5: Syntax error: "do" unexpected
```

bash output:
```
./test.sh: line 4: while: command not found
./test.sh: line 5: syntax error near unexpected token `do'
./test.sh: line 5: `do'
```

zsh output:
```
a
b
c
```

**Comment by pabs3** [comment]
Confirmed, thanks for pointing that out.

I guess the proposed check could detect the `cat | while` construct and
change the output, or just not emit `SC2002` for that construct.

-- 
bye,
pabs

https://bonedaddy.net/pabs3/

**Comment by flexibeast** [comment]
> or just not emit `SC2002` for that construct.

That would certainly work for me, since shellcheck emitted SC2002 while i was testing the possibility of moving from a `for F in` construct to a `while read` construct in [my epub-create script](https://github.com/flexibeast/epub-create/blob/main/epub-create), which is intended to be POSIX-compliant. (Although the `while read` approach also creates descriptor issues when invoking an editor within the relevant loop.)
